plugins/catalog-backend/src/service/response/write.ts

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Response } from 'express';
+import { EntitiesResponseItems } from '../../catalog/types';
+import { JsonValue } from '@backstage/types';
+import { NotFoundError } from '@backstage/errors';
+
+const JSON_CONTENT_TYPE = 'application/json; charset=utf-8';
+
+export function writeSingleEntityResponse(
+  res: Response,
+  response: EntitiesResponseItems,
+  notFoundMessage: string,
+) {
+  if (response.type === 'object') {
+    if (!response.entities[0]) {
+      throw new NotFoundError(notFoundMessage);
+    }
+
+    res.json(response.entities[0]);
+  } else {
+    if (!response.entities[0]) {
+      throw new NotFoundError(notFoundMessage);
+    }
+
+    res.setHeader('Content-Type', JSON_CONTENT_TYPE);
+    res.end(response.entities[0]);
+  }
+}
+
+export async function writeEntitiesResponse(
+  res: Response,
+  response: EntitiesResponseItems,
+  responseWrapper?: (entities: JsonValue) => JsonValue,
+) {
+  if (response.type === 'object') {
+    res.json(
+      responseWrapper
+        ? responseWrapper?.(response.entities)
+        : response.entities,
+    );
+    return;
+  }
+
+  res.setHeader('Content-Type', JSON_CONTENT_TYPE);
+
+  // responseWrapper allows the caller to render the entities within an object
+  let trailing = '';
+  if (responseWrapper) {
+    const marker = `__MARKER_${Math.random().toString(36).slice(2, 10)}__`;
+    const wrapped = JSON.stringify(responseWrapper(marker));
+    const parts = wrapped.split(`"${marker}"`);
+    if (parts.length !== 2) {
+      throw new Error(
+        `Entity items response was incorrectly wrapped into ${parts.length} different parts`,
+      );
+    }
+    res.write(parts[0], 'utf8');
+    trailing = parts[1];
+  }
+
+  let first = true;
+  for (const entity of response.entities) {
+    const prefix = first ? '[' : ',';
+    first = false;
+
+    const needsDrain = !res.write(prefix + entity, 'utf8');
+    if (needsDrain) {
+      await new Promise<void>(resolve => {
+        const cont = () => {
+          res.off('drain', cont);
+          res.off('close', cont);
+          resolve();
+        };
+        res.on('drain', cont);
+        res.on('close', cont);
```

Comment:
as mentioned elsewhere - do we actually want to retain the old behavior of handling closure with throwing and aborting immediately?

Reply:
Refactored to properly close. Didn't bother figuring out a test for it at this point but did manual verification to make sure an interrupted request is handled correctly.